### PR TITLE
Setup stripe api #40

### PR DESCRIPTION
**Pull request by natey37**
### What this PR does (required):
- Hook up Stripe functionality 
- Set up frontend UI to save credit cards as payment options
- Set up API calls to retrieve users payment options, default option, and to create a new payment option
- Set up routing for Stripe controller, create back end routes:
- POST to create a new customer as well as save their payment option
- GET to all of the customers payment options
- PATCH to set a card as the default payment option
https://github.com/hatchways/team-spearmint/issues/40
https://github.com/hatchways/team-spearmint/issues/22
https://github.com/hatchways/team-spearmint/issues/34

### Screenshots / Videos (front-end only):
Screen Shot 2022-02-22 at 1 17 50 PM
Screen Shot 2022-02-22 at 1 18 03 PM
Screen Shot 2022-02-22 at 1 18 11 PM

### Any information needed to test this feature (required):
- Need stripe API keys to test feature, message me in slack for keys!

### Any issues with the current functionality (optional):
- Should probably add a route to delete a payment?
- Also I think this PR is 3 different tickets...let me know if I should create multiple PRs and cherry pick commits?

### CODE REVIEW

File: client/src/components/Stripe/CheckoutForm.tsx [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+      return;
+    }
+
+    const newPayment = await createPayment();
+
+    if (!newPayment) {
+      updateSnackBarMessage('There was an error setting up payment!');
+    } else {
+      console.log(newPayment);
+      console.log(newPayment.clientSecret);
```
Comment (ezeilo-su) [inline client/src/components/Stripe/CheckoutForm.tsx:34]: Let's remove these console logs, please.

File: client/src/components/Stripe/CheckoutForm.tsx [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            email: newPayment.billingDetails.email,
+          },
+        },
+      });
+
+      if (error) {
+        alert(error);
+      } else if (setupIntent) {
+        updateSnackBarMessage('Successfully created new payment method');
+        console.log(setupIntent);
```
Comment (ezeilo-su) [inline client/src/components/Stripe/CheckoutForm.tsx:49]: Remove console log

File: client/src/components/Stripe/CheckoutForm.tsx [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+//   return (
+//     <form id="payment-form" onSubmit={handleSubmit}>
+//       <PaymentElement id="payment-element" />
+//       <button disabled={isLoading || !stripe || !elements} id="submit">
+//         <span id="button-text">{isLoading ? <div className="spinner" id="spinner"></div> : 'Pay now'}</span>
+//       </button>
+//       {/* Show any error or success messages */}
+//       {message && <div id="payment-message">{message}</div>}
+//     </form>
+//   );
```
Comment (ezeilo-su) [inline client/src/components/Stripe/CheckoutForm.tsx:149]: Please let's remove comments/unused code to keep our code clean

File: client/src/helpers/APICalls/createPayment.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import { FetchOptions } from './../../interface/FetchOptions';
+
+export const createPayment = async () => {
+  const fetchOptions: FetchOptions = {
+    method: 'POST',
+    // headers: { 'Content-Type': 'application/json' },
+    // body: JSON.stringify({ id: profileId }),
```
Comment (ezeilo-su) [inline client/src/helpers/APICalls/createPayment.ts:7]: Let's remove

File: client/src/pages/Settings/Payment/Payment.tsx [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  const handleNextClick = () => {
+    const next = currentIndex + 2;
+    if (next > paymentMethods.length - 1) {
+      setCurrentIndex(0);
+    } else {
+      setCurrentIndex((prev) => prev + 2);
+    }
+  };
+  console.log(paymentMethods);
+  console.log(defaultPaymentMethod);
```
Comment (ezeilo-su) [inline client/src/pages/Settings/Payment/Payment.tsx:77]: Remove console logs

File: client/src/pages/Settings/Payment/Payment.tsx [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+              } else if (brand === 'mastercard') {
+                return mastercardLogo;
+              }
+            };
+
+            return (
+              <>
+                <Card className={classes.paymentCard}>
+                  <Box display="flex" flexDirection="row" justifyContent="space-between" alignItems="center">
+                    <img className={classes.cardLogo} src={logo()}></img>
```
Comment (ezeilo-su) [inline client/src/pages/Settings/Payment/Payment.tsx:102]: This `img` tag is missing an `alt` attribute. Please kindly add one :smiley:

File: server/controllers/stripe.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+      throw new Error("Profile could not be updated with stripe customer Id")
+    } else {
+      return profile
+    }
+  }
+}
+
+// @route POST "/payment/add-payment"
+// @desc create a stripe customer and add a card payment to that customer
+// @access private
```
Comment (ezeilo-su) [inline server/controllers/stripe.js:28]: Good documentation

File: server/controllers/stripe.js [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+      }
+    })
+  }
+}
+
+// @route GET "/payment/all-payment-methods"
+// @desc Get all payment methods associated with an account
+// @access private
+exports.getPaymentMethods = asyncHandler(async (req, res, next) => {
+  console.log(req.user.id)
```
Comment (ezeilo-su) [inline server/controllers/stripe.js:87]: Lt's get rid of this one

File: server/controllers/stripe.js [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+          res.status(200).send({ paymentMethods: allPaymentMethods, defaultPaymentMethod: defaultPaymentMethod})
+      }
+    }
+  }
+})
+
+//@route PATCH "/payment/set-default/:paymentId"
+//@desc Update default payment method
+//@access private
```
Comment (ezeilo-su) [inline server/controllers/stripe.js:125]: :+1:

**Review by ezeilo-su — CHANGES_REQUESTED** [review]
Hey @natey37, good work with this one. Kindly have a look at my comments and address them.
Happy coding!

**Review by ezeilo-su — COMMENTED** [review]
Good work making the requested changes :+1:

File: server/controllers/stripe.js [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    name: profile.name,
+    address: profile.address,
+    email: email,
+  });
+  if(!customer){
+    throw new Error("Unable to create a customer, please try again!")
+  } else {
+    profile.customerId = customer.id
+
+    let savedProfile = await profile.save()
```
Comment (rajivtitus) [inline server/controllers/stripe.js:17]:
```suggestion
    const savedProfile = await profile.save()
```
I don't see us reassigning this variable

**Review by rajivtitus — APPROVED** [review]
Good work with this feature! 👍🏽

**Pull request merged**